torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -95,16 +97,58 @@ class FullyShardedDataParallel(nn.Module):
             params and grads to be on same device to work with optimizer. This
             API is subject to change. Default is ``None`` in which case there
             will be no offloading.
+        auto_wrap_policy: (Optional [callable]):
+            A callable specifying a policy to recursively wrap layers with FSDP.
+            Note that this policy currently will only apply to child modules of
+            the passed in module. The overall module is always FSDP because we
+            call the constructor.
     """
 
     def __init__(
         self,
         module: nn.Module,
         process_group: Optional[ProcessGroup] = None,
-        cpu_offload: Optional[CPUOffload] = None
+        cpu_offload: Optional[CPUOffload] = None,
+        fsdp_auto_wrap_policy: Optional[callable] = None,
     ):
         torch._C._log_api_usage_once("torch.distributed.fsdp")
         super().__init__()
+        # if fsdp_auto_wrap_policy is specified, submodules should not be
+        # already wrapped, otherwise we'd attempt to double wrap them resulting
+        # in errors.
+        if fsdp_auto_wrap_policy is not None:
+            num_params = sum([p.numel() for p in module.parameters()])
+            self._check_wrapped(
+                module,
+                check_fn=lambda mod: not isinstance(mod, FullyShardedDataParallel),
+                err_fn=lambda mod: f"Expected {mod} to NOT be FullyShardedDataParallel if auto_wrap is enabled.",
+            )
+            # TODO: refactor recursive_wrap so that it is not dependent on
+            # ConfigAutoWrap.
+            config_auto_wrap = ConfigAutoWrap(auto_wrap_policy=fsdp_auto_wrap_policy, wrapper_cls=FullyShardedDataParallel)
+            with config_auto_wrap:
+                assert ConfigAutoWrap.in_autowrap_context
+                assert ConfigAutoWrap.wrapper_cls == FullyShardedDataParallel
+                assert ConfigAutoWrap.auto_wrap_policy == fsdp_auto_wrap_policy
+                wrapped_module, total_wrapped_params = ConfigAutoWrap.recursive_wrap(
+                    module,
+                    auto_wrap_policy=fsdp_auto_wrap_policy,
+                    # Note that we have the recursive_wrap skip wrapping for
+                    # the outermost (this) module otherwise it will result in a
+                    # double-wrap causing issues.
+                    only_wrap_children=True,
```

Comment:
it is OK to wrap outermost module here?  we can check after recursive_wrap(): if total_wrapped_params== num_params, it can early return in line 148?

Reply:
Discussed offline with @zhaojuanmao, agreed that we will go with the current approach as returning early will still result in a double wrap, will `self.module` will become the true FSDP instance, but the user gets back the `self` instance. Basically we need to ensure that the remainder of the ctor logic runs for the root module.